storage: distribute source work across all timely workers

aljoscha opened the pull request:
Unborked version of #14292 

### Motivation

Currently, only one worker in a `storaged` is doing the actual ingestion/reclocking work for a source. This PR changes that to use all the available timely workers, if possible. Parallelism is still bounded by the number of "partitions".

This splits the monolithic `create_raw_source` operator into three operators:
 - `source_reader` (parallelism = num workers): does the actual reading from the external system and sends along un-timestamped batches and summaries about source uppers
 - `remap` (parallelism = 1): takes in all the summaries about source uppers and periodically mints new timestamps/offsets into the remap shard. This sends along any update that have been written to persist.
 - `reclock` (parallelism = num workers): Takes in batches and updates from the `remap` operator. Uses those updates to reclock batches and send them along.

This introduces the new `ReclockFollower` which can ingest updates emitted by a `ReclockOperator` and reclock batches. This can be seen as a read-only mirror of the `ReclockOperator`. This could in theory also directly read from persist, but I chose to actively push along updates for now.

### Tips for reviewer

There are still some `WIP`/`TODO` comments.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -439,6 +447,12 @@ pub trait SourceReader {
+    /// need these to compute a "global" source upper, when determining
+    /// completeness of a timestamp.
+    // WIP: Maybe find a better name?
+    fn unconsumed_partitions(&self) -> Vec<PartitionId>;
```
I think `-> &[PartitionId]` would be nicer!

as for naming, I think `known_uncovered_partitions` might be better but its a hard one as its a very niche concept

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -497,6 +491,7 @@ pub enum SourceMessageType<Key, Value, Diff> {
 
 /// Source-agnostic wrapper for messages. Each source must implement a
 /// conversion to Message.
+#[derive(Clone, Serialize, Deserialize)]
```
clone for `: timely::Data` right? why the serialize and deserialize? is that just for an imagined world with exchanging across processes?

guswynn replied:
wait, we have the rc wrapping, is this clone still required?

aljoscha replied:
oops, this is leftover from an in-progress impl, removing!

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -543,16 +538,34 @@ impl fmt::Debug for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>, ()> {
+    // this as a follow-up issue because it's a bigger thing that breaks with
+    // the current behaviour.
+    non_definite_errors: Vec<SourceError>,
+    source_upper: HashMap<PartitionId, MzOffset>,
```
probably `new_source_upper` is a better name, as we are communicating a new, non-remapped upper

petrosagg replied:
I'm risking bikeshedding but I think we shouldn't change the name based on a process (remapping) that happens later and is unrelated to a source producing batches

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -543,16 +538,34 @@ impl fmt::Debug for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>, ()> {
+    source_upper: HashMap<PartitionId, MzOffset>,
+}
+
+/// A batch of messages from a raw source, along with the current upper.
```
I think this is copy-pasta'd

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -543,16 +538,34 @@ impl fmt::Debug for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>, ()> {
+/// A batch of messages from a raw source, along with the current upper.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct SourceUpperSummary {
+    source_upper: HashMap<PartitionId, MzOffset>,
```
same as above

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -789,6 +836,19 @@ where
             };
 
             let mut source_upper = timestamper.source_upper_at(upper_ts.saturating_sub(1));
```
Is this `ReclockOperator` only used for this one `source_upper` call? if so, perhaps it can be passed in instead

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
+            // Emit batches more frequently than we mint new timestamps. We're
+            // hoping that most of the batches that we emit will end up making
+            // it into the freshly minted bindings when remap_operator ticks.
+            let tick_interval = timestamp_frequency / 5;
```
I would note that if data is coming in fast enough, we will likely beat this interval

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
+
+                let now = Instant::now();
+
+                if let Some(sleep_time) = sleep_time {
+                    // Sleep the min of the suggested sleep time and what time
+                    // remains until the next tick should happen.
+                    let sleep_time =
+                        std::cmp::min(sleep_time, tick_interval + (last_tick_time - now));
+
+                    tokio::time::sleep(sleep_time).await;
```
you may prefer handling this with a tokio `Interval` with a chosen missed tick behavior: https://docs.rs/tokio/latest/tokio/time/struct.Interval.html#method.set_missed_tick_behavior, to simplify some of this manual logic handling

aljoscha replied:
Can this be done without a `select!`? I thought it could not, and most of the complexity in this part comes from getting rid of that `select!`. This is in flux, Petros also has thoughts about this that he mentioned to me on slack.

guswynn replied:
it should be able to be awaited like the `sleep` here, `interval.tick()`

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
+                // We forward only the partitions that we are responsible for to
+                // the remap operator.
+                let source_upper_summary = SourceUpperSummary {
+                    source_upper: source_upper.clone(),
```
where is this `source_upper` filtered to the specific partitions?

aljoscha replied:
It's not filtered because the source only emits an upper containing the partitions that it is responsible for. We could assert that here, though.

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
+                    source_upper: source_upper.clone(),
+                };
+
+                // Pull the upper to `max` for partitions that we are not
+                // responsible for. That way, the downstream reclock operator
+                // can correcly decide when a reclocked timestamp is closed.
+                // We basically take those partitions "out of the calculation".
+                let mut extended_source_upper = source_upper.clone();
+                extended_source_upper.extend(
+                    unconsumed_partitions
+                        .into_iter()
+                        .map(|pid| (pid, MzOffset { offset: u64::MAX })),
+                );
```
this seems pretty subtle, what happens if one worker receives information about a new partition it is responsible before another worker that isnt does? The remap operator may receive data for a new partition, but the `unconsumed_partitions` for the second worker will be wrong? What kind of error can that cause?

aljoscha replied:
That wouldn't be a problem, and I would expect this to happen. The `SourceMessageBatch` is sent along a `Pipeline` edge to the `reclock_operator` on this worker. This one will then have to wait on binding information that includes that partition, which will happen once the `remap_operator` learns about it from all the `source_reader_operator`s.

guswynn replied:
we should leave a comment explaining as much!

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
+                        source_id: id,
+                        error: e.inner,
+                    })
+                    .collect_vec();
```
I wonder how much additional allocation we are doing here; This may represent additional slowdown for the pg source, but should be a relatively fixed cost per batch, i think

guswynn commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
+    ((batch_stream, summary_stream), Some(capability))
+}
+
+/// Mints new contents for the remap shard based on summaries about the source
+/// upper it receives from the raw reader operators.
```
```suggestion
/// A _single-instance_ operator that mints new contents for the remap shard based on summaries about the source
/// upper it receives from the raw reader operators.
```

aljoscha replied:
Added something to that effect to the rustdoc but I think it's better to not use `single-instance operator` because I don't know how clear the meaning of this is to most folks.

guswynn reviewed (commented):
Left some initial review. There is nothing that stands out to me as concerning other than the management of `unconsumed_partitions`, which i left a comment on

petrosagg commented on src/storage/src/source/mod.rs:
```diff
@@ -806,169 +866,640 @@ where
-                                            healthchecker.update_status(update).await;
-                                        }
-                                    }
+                let item = source_reader.get_next_message();
```
Hm, we can't call `get_next_message` here because its default implementation always returns `Pending` (I should have probably made it panic instead) and doesn't delegate to `SourceReader::next` meaning that any source that only implements `next` will not work. We don't have any source that does this at the moment but it's a step back if we still want to deprecate `get_next_message`.

aljoscha commented:
@guswynn & @petrosagg I pushed some changes, with an alternative approach for learning about the "unconsumed partitions" (yes, bad name, but this will go away with some follow-up work that Petros has planned)

aljoscha commented:
I had a great idea that is obvious in hindsight! (it's sparked by a thing Petros said recently) I've renamed `UnconsumedPartitions` to `DropPartitionCapabilities`, and I've extended the rustdoc of `SourceReader` to explain the contract.

I've also rebased this on `main` and added a bunch of "pure code movement" commits on top of my main commit. So when reviewing you should only review that main commit and not worry about the code movement.

petrosagg commented on src/storage/src/source/mod.rs:
```diff
@@ -850,125 +965,553 @@ where
+                });
+
+                if let Err(wait_time) = timestamper.next_mint_timestamp() {
+                    activator.activate_after(wait_time);
```
It's strange to see manual activations in an async operator. Does something break if we do `tokio::time::sleep(wait_time).await;` here?

aljoscha replied:
Will do! I did it because I didn't want to block reading of inputs while we wait.

aljoscha replied:
Ah: if we don't worry about not reading from the inputs, I can just remove the `if` entirely. The `timestamper.mint()` call just below this will `await` until it gets the next mint timestamp. WDYT?

petrosagg replied:
that makes sense to me, yeah!

petrosagg commented on src/storage/src/source/mod.rs:
```diff
@@ -850,125 +965,553 @@ where
+                    Err((pid, offset)) => panic!("failed to reclock {} @ {}", pid, offset),
+                };
+
+                if let Some(reclocked) = reclocked {
```
This also looks like a desugared match statement

aljoscha replied:
:+1:

petrosagg commented on src/storage/src/source/mod.rs:
```diff
@@ -850,125 +965,553 @@ where
+            );
+
+            // WIP: Be smarter about this.
+            while let Some(untimestamped_batch) = untimestamped_batches.first_mut() {
```
We should probably use `VecDeque` for `untimestamped_batches` to have a cheap `VecDeque::pop_front()` operation. `Vec` will copy all elements to slide everything back one position when we do `remove(0)`

aljoscha replied:
:+1:

petrosagg commented on src/storage/src/source/mod.rs:
```diff
@@ -982,15 +1525,34 @@ where
+
+            // It can happen that our view of the global source_upper is not yet
+            // up to date with what the ReclockOperator thinks. Ignore that for
+            // now.
```
"Ignore that for now" means that this is something to be fixed in the code in the future or that the function will ignore it for now and it will do the correct thing in a future execution (and therefore the code needs no changes)?

aljoscha replied:
Updated the comment. The operator will eventually learn about new frontiers in future invocations.

petrosagg commented on src/storage/src/source/reclock.rs:
```diff
@@ -287,6 +445,14 @@ impl ReclockOperator {
+
+    /// Returns the current contents of the remap trace. Suitable for
+    /// bootstrapping a `ReclockListener`.
+    pub fn remap_trace(&self) -> Vec<(PartitionId, Vec<(u64, MzOffset)>)> {
```
See comment above for using the cloned hashmap directly

petrosagg commented on src/storage/src/source/reclock.rs:
```diff
@@ -30,10 +30,193 @@ use mz_persist_client::read::{Listen, ListenEvent, ReadHandle};
+    /// Pushes new trace updates into this [`ReclockFollower`].
+    pub fn push_trace_updates(
+        &mut self,
+        updates: impl Iterator<Item = (PartitionId, Vec<(u64, MzOffset)>)>,
```
The trait expected by`for` loops is `IntoIterator` so we should change that to `impl IntoIterator` to make life easier for callers.

Having the type like that will allow a caller to pass a `HashMap<PartitionId, Vec<(u64, MzOffset)>>` directly so we can get rid of all the `trace_updates.into_iter().collect()`calls below

aljoscha replied:
:+1:

petrosagg commented on src/storage/src/source/reclock.rs:
```diff
@@ -295,7 +461,13 @@ impl ReclockOperator {
+    pub async fn mint<P: Borrow<PartitionId>>(
+        &mut self,
+        source_frontier: &HashMap<P, MzOffset>,
+    ) -> Vec<(PartitionId, Vec<(u64, MzOffset)>)> {
```
Can we use the type alias `Timestamp` in all places where a `u64` is used as a timestamp?

aljoscha replied:
This has been there since the big bang, but will change. :sweat_smile:

petrosagg reviewed (commented):
Looks really good! I went through the changes and the handling of unconsumed partitions and everything checks out. I've left a few comments, mostly rust-y stuff.

The only big item that seems still pending is re-enabling the reclocking unit tests

aljoscha commented on src/storage/src/source/mod.rs:
```diff
@@ -543,16 +557,40 @@ impl fmt::Debug for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>, ()> {
+}
+
+/// The source upper at the time of emitting a batch. This contains only the
+/// partitions that a given source reader operator is responsible for, so a
```
It is true, because the `remap_operator` does not yet do proper "mutable frontier" accounting the way you suggested. So if one operator would emit a `(pid, MAX)` here, that would "drop" that pid in the remap operator.

aljoscha commented on src/storage/src/source/mod.rs:
```diff
@@ -850,125 +945,553 @@ where
+
+            while let Poll::Ready(Some(update)) = source_reader.as_mut().poll_next(&mut context) {
+                let (messages, non_definite_errors, unconsumed_partitions, source_upper) =
+                    if let Some(update) = update {
```
:+1:

aljoscha commented on src/storage/src/source/mod.rs:
```diff
@@ -806,26 +891,33 @@ where
             loop {
                 // TODO(guswyn): move lots of this out of the macro so rustfmt works better
+                // TODO(aljoscha): Get rid of the async stream here. To ease my
+                // cancel-safety anxiety.
```
:+1:

guswynn commented on src/storage/src/source/generator.rs:
```diff
@@ -40,6 +40,14 @@ pub struct LoadGeneratorSourceReader {
+    // that is is not consuming from the one [`PartitionId:None`] partition.
+    // Before it can return a [`NextMessage::Finished`]. This is keeping track
+    // of that.
+    reported_unconsumed_partitions: bool,
```
this never seems to be set to `true`?

aljoscha replied:
Oh boy, good catch! This wasn't caught because we never run sources with more than one timely worker. Good thing that this removes barriers to doing that. I think @philip-stoev is eager to enable multi-worker sources in tests. :sweat_smile:

guswynn commented on src/storage/src/source/reclock.rs:
```diff
@@ -523,21 +729,44 @@ mod tests {
+        (operator, follower)
+    }
+
+    async fn mint_and_follow(
+        operator: &mut ReclockOperator,
+        follower: &mut ReclockFollower,
+        source_upper: &mut HashMap<PartitionId, MzOffset>,
+    ) {
+        let trace_updates = operator.mint(source_upper).await;
+        let reclock_upper = operator
+            .reclock_frontier(&source_upper)
+            .expect("wrong source upper");
+        follower.push_trace_updates(trace_updates.into_iter());
+        follower.push_upper_update(reclock_upper);
```
love this!

guswynn reviewed (commented):
I have one tiny nit about a boolean variable left, but thats all, I defer to @petrosagg on the approval!
